server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java

```diff
@@ -0,0 +1,160 @@
+        return this.positionIncrementGap;
+    }
+
+    @Override
+    public int getOffsetGap(String field) {
+        if (this.offsetGap < 0) {
+            return super.getOffsetGap(field);
+        }
+        return this.offsetGap;
+    }
+
+    public AnalysisMode getAnalysisMode() {
+        return this.components.analysisMode();
+    }
+
+    @Override
+    protected Reader initReaderForNormalization(String fieldName, Reader reader) {
+        final AnalyzerComponents components = getComponents();
+        for (CharFilterFactory charFilter : components.getCharFilters()) {
+            reader = charFilter.normalize(reader);
+        }
+        return reader;
+    }
+
+    @Override
+    protected TokenStream normalize(String fieldName, TokenStream in) {
+        final AnalyzerComponents components = getComponents();
+        TokenStream result = in;
+        for (TokenFilterFactory filter : components.getTokenFilters()) {
+            result = filter.normalize(result);
+        }
+        return result;
+    }
+
+    private boolean shouldReload(AnalyzerComponents source) {
+        return this.components != source;
+    }
+
+    public synchronized void reload(String name,
+                                    Settings settings,
+                                    final Map<String, TokenizerFactory> tokenizers,
+                                    final Map<String, CharFilterFactory> charFilters,
+                                    final Map<String, TokenFilterFactory> tokenFilters) {
+        AnalyzerComponents components = AnalyzerComponents.createComponents(name, settings, tokenizers, charFilters, tokenFilters);
+        this.components = components;
+    }
+
+    @Override
+    public void close() {
+        storedComponents.close();
+    }
+
+    private void setStoredComponents(AnalyzerComponents components) {
+        storedComponents.set(components);
+    }
+
+    private AnalyzerComponents getStoredComponents() {
+        return storedComponents.get();
+    }
+
+    @Override
+    protected TokenStreamComponents createComponents(String fieldName) {
+        final AnalyzerComponents components = getStoredComponents();
+        Tokenizer tokenizer = components.getTokenizerFactory().create();
+        TokenStream tokenStream = tokenizer;
+        for (TokenFilterFactory tokenFilter : components.getTokenFilters()) {
+            tokenStream = tokenFilter.create(tokenStream);
+        }
+        return new TokenStreamComponents(tokenizer, tokenStream);
+    }
+
+    @Override
+    protected Reader initReader(String fieldName, Reader reader) {
+        final AnalyzerComponents components = getStoredComponents();
+        if (components.getCharFilters() != null && components.getCharFilters().length > 0) {
+            for (CharFilterFactory charFilter : components.getCharFilters()) {
+                reader = charFilter.create(reader);
+            }
+        }
+        return reader;
+    }
+}
```

Comment:
Summary of comments from #42669:

Alan expressed the ides to have ReloadableCustomAnalyzer extend CustomAnalyzer, which is something I originally did in the first WIP PR (#40782) which Jim was somewhat against. He suggested pulling out a common base class. 

My take on this, that I implemented in a follow up commit here, is to make both Analyzers implement a common interface that allows uniform access to the AnalyzerComponents without having to change the CustomAnalyzer API much.